library/std/src/process.rs

```diff
@@ -1404,6 +1404,14 @@ impl From<fs::File> for Stdio {
 /// For proper error reporting of failed processes, print the value of `ExitStatus` or
 /// `ExitStatusError` using their implementations of [`Display`](crate::fmt::Display).
 ///
+/// # Differences from `ExitStatus`
+///
+/// The primary difference between [`ExitStatus`] and `ExitCode` is that
+/// `ExitStatus` can represent either a exit code or an exit signal, such as
+/// "killed by signal", whereas `ExitCode` can only represent exit codes.
+/// `Termination` uses `ExitCode` instead of `ExitStatus` because Rust does not
+/// support normal termination (aka return from main) with a signal.
```

Comment:
I think this may lead people to believe that e.g. SIGINT from Ctrl+C are UB in Rust, which (AFAIK) is not true. It also seems like we *could* support ending as-if a signal had occurred, right?

Reply:
I believe we'd have to switch the handler to default and then self-send signal. Which is not guaranteed to be delivered. Sounds very tricky and scary.

Reply:
> I think this may lead people to believe that e.g. SIGINT from Ctrl+C are UB in Rust, which (AFAIK) is not true.

I'm not really sure how best to avoid this misinterpretation. I think part of the issue here is I don't have a clear mental model for exactly what signals should be used for or what they represent. In my mind they're akin to an externally initiated abort,  The best idea I can think of is to reword the last sentence a bit to make it more focused on this specific interface rather than the task it represents. Something like:

> `Termination` uses `ExitCode` instead of `ExitStatus` because `Termination` is meant to represent normal (aka graceful) termination in a cross platform manner, and signals are not supported on all Tier 1 platforms supported by Rust. Additionally, where they are used for process termination, signals are generally used to abruptly kill a process from a separate process running within the same OS, not as a return value from `main`.

> > It also seems like we _could_ support ending as-if a signal had occurred, right?
> 
> I believe we'd have to switch the handler to default and then self-send signal. Which is not guaranteed to be delivered. Sounds very tricky and scary.

I'm not an expert on signals by any means, but I think it is also the case that signal handling is fairly platform specific. A quick glance at the various ExitStatusExt traits show that extracting signals isn't part of the cross platform interface for `ExitStatus`. It's available on unix via the extension trait and it's not available on windows at all.

Even if we could plausibly introduce an interface for exiting via a signal I suspect that what we would want is closer to `std::process::exit` than it would be to `Termination::report`.

Reply:
I think it might help to reframe this as something more positive (rather than focusing on the "lack" of support for signals). Maybe something like:

> `ExitCode` is intended for terminating the currently running process, via the `Termination` trait, in contrast to `ExitStatus`, which represents the termination of a child process. These APIs are separate due to platform compatibility differences and their expected usage; it is not generally possible to exactly reproduce an ExitStatus from a child for the current process after the fact.

This does make me think that maybe naming them ExitCode and ExitStatus is a bit confusing and we should hunt for a different name, but my "best" ideas are ... bad, so I think we shouldn't block stabilization on this. It feels a little weird that we have Termination with `report(self) -> ExitCode` rather than `exit(ExitCode)`, and some From impls, but I guess that ship has somewhat sailed given how late in the process things are and the partial exposure of Termination to stable code -- though I *think* we could get away with a change if main basically returned `impl Into<ExitCode>` which is effectively a less-fancy name for Termination.

Reply:
> This does make me think that maybe naming them ExitCode and ExitStatus is a bit confusing and we should hunt for a different name, but my "best" ideas are ... bad, so I think we shouldn't block stabilization on this. It feels a little weird that we have Termination with `report(self) -> ExitCode` rather than `exit(ExitCode)`, and some From impls, but I guess that ship has somewhat sailed given how late in the process things are and the partial exposure of Termination to stable code -- though I _think_ we could get away with a change if main basically returned `impl Into<ExitCode>` which is effectively a less-fancy name for Termination.

I don't think `Into` is the appropriate trait here, since the usage of `Termination::report` also heavily implies side effects during the conversion aka printing the error to stderr. Expecting `Into` impls to call `eprintln!` feels like an abuse of the trait. Also, I'm always incredibly worried about running into overlap issues when using the `From`/`Into` traits, so I'm fairly worried about what restrictions we might unknowingly put upon ourselves if we tried to piggyback on existing traits for this.

Reply:
> I'm not an expert on signals by any means

I'm not sure if I should call myself "expert" but I did work with them quite a bit and they are **very** ugly interface. Ideally avoid by default. There are some reasonable tricks to use them safely and sanely but self sending doesn't seem that useful. (I did actually see it in practice but can't remember why guys didn't just execute the signal handler directly.)

> Even if we could plausibly introduce an interface for exiting via a signal I suspect that what we would want is closer to `std::process::exit` than it would be to Termination::report`.

Agreed

@Mark-Simulacrum nice description. Can't figure out a better name either.

Reply:
Ah, right, I had forgotten about the side-effects. Seems good.